File: atcoder/abc436/d.py
from collections import deque, defaultdict

H, W = map(int, input().split())
S = [input() for _ in range(H)]

pos = defaultdict(list)
for r in range(H):
    for c in range(W):
        if S[r][c].isalpha():
            pos[S[r][c]].append((r, c))

inf = 10**9
dis = [[inf for _ in range(W)] for _ in range(H)]
que = deque()

dis[0][0] = 0
que.append((0, 0))

while len(que) > 0:
    r, c = que.popleft()

    # walk
    for nr, nc in [(r + 1, c), (r - 1, c), (r, c + 1), (r, c - 1)]:
        if nr < 0 or nr >= H or nc < 0 or nc >= W:
            continue
        if S[nr][nc] != "#" and dis[nr][nc] == inf:
            dis[nr][nc] = dis[r][c] + 1
            que.append((nr, nc))

    # warp
    for nr, nc in pos[S[r][c]]:
        if dis[nr][nc] == inf:
            dis[nr][nc] = dis[r][c] + 1
            que.append((nr, nc))
    pos[S[r][c]].clear()  # !!!!!!!!!!

if dis[-1][-1] == inf:
    print(-1)
else:
    print(dis[-1][-1])

File: atcoder/abc436/d.rs
#![allow(unused)]

use std::collections::VecDeque;

use proconio::{input, marker::Chars};

fn main() {
    input! {
        h: usize,
        w: usize,
        s: [Chars; h],
    }

    let mut adj = vec![vec![]; h * w + 26];

    for r in 0..h {
        for c in 0..w {
            // walk
            if s[r][c] != '#' {
                let u = r * w + c;
                for (dr, dc) in [(0, 1), (0, -1), (1, 0), (-1, 0)] {
                    let nr = r.checked_add_signed(dr).unwrap_or(h);
                    let nc = c.checked_add_signed(dc).unwrap_or(w);
                    if nr < h && nc < w && s[nr][nc] != '#' {
                        let v = nr * w + nc;
                        adj[u].push((v, 1));
                    }
                }
            }
            // warp
            if s[r][c].is_ascii_lowercase() {
                let u = r * w + c;
                let z = h * w + s[r][c] as usize - 'a' as usize;
                adj[u].push((z, 1));
                adj[z].push((u, 0));
            }
        }
    }

    let inf = 1_000_000_000;
    let (dis, _) = bfs01(&adj, 0, inf);

    let ans = dis[(h - 1) * w + (w - 1)];
    if ans != inf {
        println!("{ans}");
    } else {
        println!("-1");
    }
}

fn bfs01<T>(adj: &Vec<Vec<(usize, T)>>, s: usize, inf: T) -> (Vec<T>, Vec<usize>)
where
    T: std::ops::Add<Output = T> + Copy + Default + Eq + Ord,
{
    let n = adj.len();
    let mut que = VecDeque::new();
    let mut dis = vec![inf; n];
    let mut par = vec![!0; n];

    dis[s] = T::default();
    par[s] = s;
    que.push_back((s, dis[s]));

    while let Some((u, d)) = que.pop_front() {
        if d > dis[u] {
            continue;
        }
        for &(v, w) in &adj[u] {
            let new_d = dis[u] + w;
            if new_d < dis[v] {
                dis[v] = new_d;
                par[v] = u;
                if w == T::default() {
                    que.push_front((v, dis[v]));
                } else {
                    que.push_back((v, dis[v]));
                }
            }
        }
    }

    (dis, par)
}

fn join<T: ToString>(arr: &[T], sep: &str) -> String {
    arr.iter()
        .map(|x| x.to_string())
        .collect::<Vec<String>>()
        .join(sep)
}
